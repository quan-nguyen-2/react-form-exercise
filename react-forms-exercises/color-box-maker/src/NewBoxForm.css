input {
    width: 420px;
    height: 25px;
    font-size: 1.5rem;
    border-radius: 5px;
    border: 1px solid gray;
}

.NewBoxForm-btn {
    background-color: seagreen;
    color: white;
    cursor: pointer;
    border: none;
    font-size: 1.5rem;
    border-radius: 5px;
    height: 2.5rem;
}


label {
    font-size:1.5rem;
    margin-right: 1rem;
}

form {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
}

.field {
    margin-bottom: 1rem;
    display: flex;
    justify-content: end;
}

h3 {
    font-size: 3rem;
}

span {
    margin-bottom: 1rem;
}