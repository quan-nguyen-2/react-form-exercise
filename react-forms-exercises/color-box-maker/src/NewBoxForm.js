import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import "./NewBoxForm.css";

const NewBoxForm = ({ addBox }) => {
  const INITIAL_STATE = {
    width: "",
    height: "",
    color: "",
  };

  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleChange = (e) => {
    //destructure name and value from inputs
    const { name, value } = e.target;
    //update state of formData on every change to any form input
    setFormData((formData) => ({ ...formData, [name]: value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    //use parent component function to add box to parent's 'boxes' state
    addBox({ ...formData, id: uuid() });
    //reset form state to empty
    setFormData(INITIAL_STATE);
  };

  return (
    <form className="NewBoxForm" onSubmit={handleSubmit}>
      <div className="field">
        <label htmlFor="width">Width</label>
        <input
          type="number"
          id="width"
          name="width"
          onChange={handleChange}
          value={formData.width}
        />
      </div>
      <div className="field">
        <label htmlFor="height">Height</label>
        <input
          type="number"
          id="height"
          name="height"
          onChange={handleChange}
          value={formData.height}
        />
      </div>
      <div className="field">
        <label htmlFor="color">Color</label>
        <input
          type="text"
          id="color"
          name="color"
          onChange={handleChange}
          value={formData.color}
        />
      </div>

      <button className="NewBoxForm-btn">Add Box</button>
    </form>
  );
};

export default NewBoxForm;
