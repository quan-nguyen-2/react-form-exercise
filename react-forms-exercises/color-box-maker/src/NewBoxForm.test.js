import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

it("renders without crashing", function () {
  render(<NewBoxForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("should accept input data", function () {
  render(<NewBoxForm />);

  //select the form inputs and submit button
  const widthInput = screen.getByLabelText("Width");
  const heightInput = screen.getByLabelText("Height");
  const colorInput = screen.getByLabelText("Color");
  // const submitBtn = screen.getByText("Add Box");

  //enter data into form inputs
  fireEvent.change(widthInput, { target: { value: "7" } });
  expect(widthInput.value).toBe("7");
  fireEvent.change(heightInput, { target: { value: "7" } });
  expect(heightInput.value).toBe("7");
  fireEvent.change(colorInput, { target: { value: "indigo" } });
  expect(colorInput.value).toBe("indigo");
});

it("runs the add function on form submit", function () {
  const createMock = jest.fn();
  render(<NewBoxForm addBox={createMock} />);
  const addBtn = screen.getByText("Add Box");
  fireEvent.click(addBtn);
  expect(createMock).toHaveBeenCalled();
});
