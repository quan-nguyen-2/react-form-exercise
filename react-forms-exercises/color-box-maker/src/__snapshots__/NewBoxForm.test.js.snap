// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewBoxForm"
  >
    <div
      class="field"
    >
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
    </div>
    <div
      class="field"
    >
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
    </div>
    <div
      class="field"
    >
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
    </div>
    <button
      class="NewBoxForm-btn"
    >
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
